### PR TITLE
User registration feature

### PR DESCRIPTION
**Pull request by artemklishch**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -104,6 +104,12 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.6.0</version>
         </dependency>
+
+        <dependency>
+            <groupId>org.mvel</groupId>
+            <artifactId>mvel2</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml]: why do you need this dependency?

File: src/main/java/org/example/intro/dto/user/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.hibernate.validator.constraints.Length;
+
+@Setter
+@Getter
+@FieldMatch(field = "password", verifyField = "repeatPassword")
+public class UserRegistrationRequestDto {
+    @NotBlank(message = "Email field is mandatory")
+    @Email(message = "Should be a valid email.")
+    @Length(max = 255, message = "Email can not be over 255 characters")
+    String email;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/dto/user/UserRegistrationRequestDto.java]:
```suggestion
    private String email;
```

all fields should be private

File: src/main/java/org/example/intro/model/User.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    private Long id;
+    @Column(unique = true, nullable = false)
+    private String email;
+    @Column(nullable = false)
+    private String password;
+    @Column(name = "first_name", nullable = false)
+    private String firstName;
+    @Column(name = "last_name", nullable = false)
+    private String lastName;
+    @Column(name = "shipping_address")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/User.java]:
```suggestion
```

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            throw new RegistrationException(
+                    "Can't register user with email "
+                            + requestDto.getEmail()
+                            + ". Email is already in use."
+            );
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/UserServiceImpl.java]: use Mapper for these logic

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
+        User savedUser = userRepository.save(user);
+        return userMapper.toUserResponseDto(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        return userMapper.toUserResponseDto(user);
```

File: src/main/java/org/example/intro/validation/FieldMatchValidator.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        if (!matches) {
+            context.disableDefaultConstraintViolation();
+            context.buildConstraintViolationWithTemplate(
+                            "Fields '{field}' and '{verifyField}' must match"
+                    )
+                    .addNode(verifyField)
+                    .addConstraintViolation();
+        }
+
+        return matches;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/validation/FieldMatchValidator.java]:
```suggestion
        Object field = new BeanWrapperImpl(value).getPropertyValue(this.field);
        Object fieldMatch = new BeanWrapperImpl(value).getPropertyValue(this.verifyField);
        return Objects.equals(field, fieldMatch);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve it

File: src/main/java/org/example/intro/validation/FieldMatchValidator.java (hunk 7 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/validation/FieldMatchValidator.java]: ++, updated. Thank you!

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/UserServiceImpl.java]: ++, updated. Thank you!

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java (hunk 5 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/UserServiceImpl.java]: ++, updated. Thank you!

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/UserServiceImpl.java]: ++, updated. Thank you!

File: src/main/java/org/example/intro/model/User.java (hunk 3 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/User.java]: ++, removed

File: src/main/java/org/example/intro/dto/user/UserRegistrationRequestDto.java (hunk 2 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/user/UserRegistrationRequestDto.java]: ++, updated. Thank you!

File: pom.xml (hunk 1 above)
Comment (artemklishch) [reply pom.xml]: i maid custom @FieldMatch validation and use this dependency there. but i see you've simplified the validation method and here i remove it now. Thank you!

File: src/main/java/org/example/intro/service/UserService.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package org.example.intro.service;
+
+import org.example.intro.dto.user.UserRegistrationRequestDto;
+import org.example.intro.dto.user.UserResponseDto;
+import org.example.intro.exceptions.RegistrationException;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/UserService.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

File: src/main/java/org/example/intro/service/UserService.java (hunk 8 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/UserService.java]: ++, removed. Thank you!

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**